### DISCUSSION TITLE
Got `rpc error: code = PermissionDenied` when executing `terraform plan`

### DISCUSSION THREAD

**Discussion by TakumiHaruta** [zitadel/zitadel#5056] [Q&A] [answered]
Hi there,

I got the error below when I execute `terraform plan` to Zitadel Cloud.
Is this just a bug or am I missing some settings?

```
╷
│ Error: error while listing orgs: rpc error: code = PermissionDenied desc = No matching permissions found (AUTH-5mWD2)
│ 
│   with module.zitadel.zitadel_org.org["aaa"],
│   on ../shared/modules/zitadel/zitadel.tf line 1, in resource "zitadel_org" "org":
│    1: resource "zitadel_org" "org" {
│ 
╵
```

## How to reproduce

1. Create an instance `aaa` in Zitadel Cloud
2. Add a service account named `Terraform` in `aaa` for Terraform
3. Add `Org Owner` role of the default org to `Terraform`
4. Create a key from `Terraform` and download json file
5. Prepare the HCL file below

```
# main.tf
terraform {
  required_providers {
    zitadel = {
      source  = "zitadel/zitadel"
      version = "1.0.0-alpha.11"
    }
  }
  required_version = "~> 1.2.2"
}

provider "zitadel" {
  domain = "<id>.zitadel.cloud"
  token  = "~/zitadel/196931135370952961.json"
}

module "zitadel" {
  source = "../shared/modules/zitadel"
}
```

```
# ../shared/modules/zitadel/provider.tf

terraform {
  required_providers {
    zitadel = {
      source  = "zitadel/zitadel"
      version = "1.0.0-alpha.11"
    }
  }
}
```

```
# ../shared/modules/zitadel/zitadel.tf

resource "zitadel_org" "org" {
  for_each = var.tenant_ids
  name     = "${terraform.workspace}-${each.key}"
}

resource "zitadel_project" "project" {
  for_each               = var.tenant_ids
  name                   = "${terraform.workspace}-${each.key}"
  org_id                 = zitadel_org.org[each.key].id
  project_role_assertion = false
  project_role_check     = false
  has_project_check      = false

  depends_on = [
    zitadel_org.org
  ]
}

variable "tenant_ids" {
  type = map(map(list(string)))
  default = {
    "aaa" : {},
    "bbb" : {}
  }
}
```

6. execute `terraform init`

```
$ terraform init
Initializing modules...
- zitadel in ../shared/modules/zitadel

Initializing the backend...

Initializing provider plugins...
- Finding zitadel/zitadel versions matching "1.0.0-alpha.11"...
- Installing zitadel/zitadel v1.0.0-alpha.11...
- Installed zitadel/zitadel v1.0.0-alpha.11 (signed by a HashiCorp partner, key ID F3B2EB3D8891CFFF)

...

Terraform has been successfully initialized!

...
```

7. Execute `terraform init` -> The first creation has succeeded

```
$ terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.zitadel.zitadel_org.org["aaa"] will be created
  + resource "zitadel_org" "org" {
      + id   = (known after apply)
      + name = "default-aaa"
    }

  # module.zitadel.zitadel_org.org["bbb"] will be created
  + resource "zitadel_org" "org" {
      + id   = (known after apply)
      + name = "default-bbb"
    }

  # module.zitadel.zitadel_project.project["aaa"] will be created
  + resource "zitadel_project" "project" {
      + has_project_check      = false
      + id                     = (known after apply)
      + name                   = "default-aaa"
      + org_id                 = (known after apply)
      + project_role_assertion = false
      + project_role_check     = false
      + state                  = (known after apply)
    }

  # module.zitadel.zitadel_project.project["bbb"] will be created
  + resource "zitadel_project" "project" {
      + has_project_check      = false
      + id                     = (known after apply)
      + name                   = "default-bbb"
      + org_id                 = (known after apply)
      + project_role_assertion = false
      + project_role_check     = false
      + state                  = (known after apply)
    }

Plan: 4 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.zitadel.zitadel_org.org["bbb"]: Creating...
module.zitadel.zitadel_org.org["aaa"]: Creating...
module.zitadel.zitadel_org.org["bbb"]: Creation complete after 4s [id=196932431746433281]
module.zitadel.zitadel_org.org["aaa"]: Creation complete after 4s [id=196932431964537089]
module.zitadel.zitadel_project.project["bbb"]: Creating...
module.zitadel.zitadel_project.project["aaa"]: Creating...
module.zitadel.zitadel_project.project["aaa"]: Creation complete after 2s [id=196932434917327105]
module.zitadel.zitadel_project.project["bbb"]: Creation complete after 2s [id=196932435571638529]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
```

7. Execute `terraform plan` -> Got the permission errors

```
$ terraform plan
module.zitadel.zitadel_org.org["aaa"]: Refreshing state... [id=196932431964537089]
module.zitadel.zitadel_org.org["bbb"]: Refreshing state... [id=196932431746433281]
╷
│ Error: error while listing orgs: rpc error: code = PermissionDenied desc = No matching permissions found (AUTH-5mWD2)
│ 
│   with module.zitadel.zitadel_org.org["aaa"],
│   on ../shared/modules/zitadel/zitadel.tf line 1, in resource "zitadel_org" "org":
│    1: resource "zitadel_org" "org" {
│ 
╵
╷
│ Error: error while listing orgs: rpc error: code = PermissionDenied desc = No matching permissions found (AUTH-5mWD2)
│ 
│   with module.zitadel.zitadel_org.org["bbb"],
│   on ../shared/modules/zitadel/zitadel.tf line 1, in resource "zitadel_org" "org":
│    1: resource "zitadel_org" "org" {
│ 
```

Provisioning orgs and projects by Terraform are very important for our use cases.
We'd be glad if you could help us!

**Comment by fforootd** [comment, marked as answer]
Hey there, since you want to provision multiple orgs you need to provide higher management rights.

What happens if you use IAM_OWNER or IAM_ORG_MANAGER?

**Reply by TakumiHaruta** [reply]
@fforootd 

Where could I select `IAM_OWNER` or `IAM_ORG_MANAGER` in Zitadel cloud?
There is no such option in `Edit roles` modal.

![Screenshot 2023-01-17 at 7 45 33](https://user-images.githubusercontent.com/35373297/212776165-1e3a6b5a-a826-4cb0-9a94-cea2189cdfb3.png)

It seems to be related to instance settings but I couldn't find the section;

- https://github.com/zitadel/zitadel/blob/47ffa52f0faef028439914b006f124dc4d465751/internal/command/instance.go#L321

**Reply by TakumiHaruta** [reply]
Oh sorry I found the setting and it worked!

**Reply by fforootd** [reply]
Nice to hear that. We know the "instance" settings is a little too "hidden" and we will improve this.

**Comment by spanktar** [comment]
Greetings! I am having the same error with roughly the same configuration, the only difference is that I am not creating multiple organizations. I'm just trying to do a Hello World with the provider.

```
Terraform will perform the following actions:

  # zitadel_org.this will be created
  + resource "zitadel_org" "this" {
      + id             = (known after apply)
      + name           = "terraform-test"
      + primary_domain = (known after apply)
      + state          = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

zitadel_org.this: Creating...
╷
│ Error: rpc error: code = PermissionDenied desc = No matching permissions found (AUTH-5mWD2)
│
│   with zitadel_org.this,
│   on main.tf line 1, in resource "zitadel_org" "this":
│    1: resource "zitadel_org" "this" {
```

I'm still looking for "instance settings" with no luck (yet!)

**Reply by spanktar** [reply]
Whoa, I accidentally found it but had a very hard time figuring out how I found it. Maybe this will help someone in the future:

- First, you go to your instances in the *ADMIN* portal: https://zitadel.com/admin/instances
- Then, click the "View Details" button. You'll get to a URL like:
https://zitadel.com/admin/instances/12345678901234567890
- Then you click the "bubble" in the upper right corner, which is right next to the "Options" button. This will take you here:
https://zitadel.com/admin/instances/12345678901234567890/members
- Now click the "Edit" button (upper right) which will take you here: https://YOUR-DOMAIN-123456.us1.zitadel.cloud/ui/console/instance/members

It is on this screen you can add someone with IAM-type permissions.

I think you can get to this a different way, but I swear when I did it this way it was different (maybe not?) Anyway, start at your Instance Console:

https://terraform-test-123456.us1.zitadel.cloud/ui/console/

- Click "Default Settings" in the upper right. 
- Now click the "+" (add Manager) button next to the "Customer Portal" button in the upper right.
- Here you can add an IAM type user.
- If you want to EDIT a user, you have to click the bubbles of manages between the "Customer Portal" button and the "+" (add Manager) button on the previous screen.

LONG STORY SHORT: Here's where you want to end up:
```
https://<YOUR-DOMAIN-1234567>.us1.zitadel.cloud/ui/console/instance/members
```
